src/Microsoft.Health.Dicom.Core/Features/Workitem/WorkitemOrchestrator.cs

```diff
@@ -49,42 +49,126 @@ public class WorkitemOrchestrator : IWorkitemOrchestrator
-
-                IReadOnlyCollection<QueryTag> queryTags = await _workitemQueryTagService.GetQueryTagsAsync(cancellationToken).ConfigureAwait(false);
+                var partitionKey = _contextAccessor.RequestContext.GetPartitionKey();
+                var queryTags = await _workitemQueryTagService.GetQueryTagsAsync(cancellationToken).ConfigureAwait(false);
 
                 identifier = await _indexWorkitemStore
                     .BeginAddWorkitemAsync(partitionKey, dataset, queryTags, cancellationToken)
                     .ConfigureAwait(false);
 
                 // We have successfully created the index, store the file.
-                await StoreWorkitemBlobAsync(identifier, dataset, cancellationToken)
+                await StoreWorkitemBlobAsync(identifier, dataset, null, cancellationToken)
+                    .ConfigureAwait(false);
+
+                await _indexWorkitemStore
+                    .EndAddWorkitemAsync(identifier.PartitionKey, identifier.WorkitemKey, cancellationToken)
+                    .ConfigureAwait(false);
+            }
+            catch
+            {
+                await TryAddWorkitemCleanupAsync(identifier, cancellationToken)
+                    .ConfigureAwait(false);
+
+                throw;
+            }
+        }
+
+        /// <inheritdoc />
+        public async Task CancelWorkitemAsync(DicomDataset dataset, WorkitemMetadataStoreEntry workitemMetadata, ProcedureStepState targetProcedureStepState, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(dataset, nameof(dataset));
+            EnsureArg.IsNotNull(workitemMetadata, nameof(workitemMetadata));
+
+            if (workitemMetadata.Status != WorkitemStoreStatus.ReadWrite)
+            {
+                return;
+            }
+
+            (long CurrentWatermark, long NextWatermark)? watermarkEntry = null;
+            WorkitemInstanceIdentifier identifier = null;
+
+            try
+            {
+                watermarkEntry = await _indexWorkitemStore
+                    .GetCurrentAndNextWorkitemWatermarkAsync(workitemMetadata.WorkitemKey, cancellationToken)
+                    .ConfigureAwait(false);
+
+                // Get the workitem from blob store
+                identifier = new WorkitemInstanceIdentifier(
+                    workitemMetadata.WorkitemUid,
+                    workitemMetadata.WorkitemKey,
+                    workitemMetadata.PartitionKey,
+                    watermarkEntry.Value.CurrentWatermark);
+                var storeDicomDataset = await GetWorkitemBlobAsync(identifier, cancellationToken).ConfigureAwait(false);
+
+                // update the procedure step state
+                var updatedDicomDataset = new DicomDataset(storeDicomDataset);
+
+                // Add/Update the procedure step state in the blob
+                var targetProcedureStepStateStringValue = targetProcedureStepState.GetStringValue();
```

Comment:
Once we add update and change state, we can come back to request cancel and compose it from those two operations.

Reply:
Done